### PR TITLE
Feat: Adding Parent group name when no groups are provided

### PR DESCRIPTION
**Pull request by Nazdroth**
Feat: Adding Parent group name when no groups are provided

Allow fields with no groups to inherit their parent group visibility if it’s declared.
 
for example:

```
type UserInfo struct {
    UserPrivateInfo `groups:"private"`
    UserPublicInfo  `groups:"public"`
}
type UserPrivateInfo struct {
    Age string
}
type UserPublicInfo struct {
    ID    string
    Email string `groups:"private"`
}
```

Before adding this feature, if we request group tag “public” we get as an output:
`{0}`
And now that we use default parent groups we get:
`{ "ID": "A" } `

We notice also that it does only override childs with no groups.

### CODE REVIEW

**Comment by mweibel** [comment]
I'm a bit unsure whether this is a good idea, especially as a default. What's the reason behind this PR?
Why can't you set the groups in the struct itself?

**Comment by Segflow** [comment]
I personally think this is a good idea if a field has no group, it does make sense for it to inherit the parent anonymous struct it's in. 

We use go in our RPG game, and our codebase contains multiple structs where some fields are private and others are private.

Our structs are kinda big so we followed a similar convention as `docker/docker` and separate the structs into private and public parts, and then use struct embedding to create the whole object. Having to write the groups tag in all fields is what made us create this PR.

File: sheriff.go [hunk 1]
```diff
@@ -51,6 +54,12 @@ func Marshal(options *Options, data interface{}) (interface{}, error) {
 	v := reflect.ValueOf(data)
 	t := v.Type()
 
+	// Initialise nestedGroupsMap,
+	// TODO: this may impact the performance, find a better place for this.
+	if options.nestedGroupsMap == nil {
```
Comment (mweibel) [inline sheriff.go]: we could probably move to a `NewOptions` (or similar) function which initializes the `Options` struct for that.

Unsure if needed, so leave it as it is for the moment.

**Comment by mweibel** [comment]
Fair enough, that makes sense.

Sorry for the delay. I'll test this locally a bit hopefully within this week and then probably merge it.

**Comment by mweibel** [comment]
Oh, one thing: could you add a bit of documentation to the README?

File: README.md [hunk 2]
```diff
@@ -33,7 +33,26 @@ type GroupsExample struct {
...
+type UserInfo struct {
+    UserPrivateInfo `groups:"private"`
+    UserPublicInfo  `groups:"public"`
+}
+type UserPrivateInfo struct {
+    Age string
+}
+type UserPublicInfo struct {
+    ID    string
+    Email string`
```
Comment (mweibel) [inline README.md]: "`" too much

Comment (mweibel) [inline README.md]:
```suggestion
    Email string
```

**Comment by masseelch** [comment]
Is there a way to disable this feature? 

Its very inconsistent. Serializing a struct where some fields do not have a group tag will be ignored. But if the same struct then gets serialized within another struct, the fields suddenly appear in the serialized map. This is very confusing and unintuitiv.

**Pull request merged**